### DISCUSSION TITLE
Show all preprocessor macros recognized by CPPTools

### DISCUSSION THREAD

**Discussion by tim-hilt** [microsoft/vscode-cpptools#11334] [Q&A] [answered]
Is there a command that allows me to see all current preprocessor defines that CPPTools is aware of? CPPTools must know this somehow, judging by reading through #510

**Comment by bobbrow** [comment, marked as answer]
@tim-hilt are you looking for the default preprocessor defines coming from the command line and compiler, the ones defined by the source file and its included headers, or both?

The first is possible by running the `C/C++: Log Diagnostics` command.  The second isn't currently possible in the way you'd probably like to consume it - they all show up in the code completion suggestion list, but there's currently no way to filter that to just preprocessor defines.  The third is limited by the answer to the second option.

**Reply by tim-hilt** [reply]
That’s a very helpful response. Thank you!

**Reply by bobbrow** [reply]
Sorry, I realize I misspoke about the first option.  The `Log Diagnostics` command does not print out all of the system defined preprocessor items because for clang/gcc is the list is just too long.  We do print them out in the Output window whenever a new language server is created (i.e. for each c/cpp file you open) if you have the `C_Cpp.loggingLevel` setting set to `Debug` though.
